File: 2020/2020W10/PreppindataW10.py
from kanren import *
from kanren.core import lall
import pandas as pd

def higher(a, b, list):
    return membero((a,b), zip(list, list[1:]))

def any2(a, b, list):
    return conde([membero(a, list)],[membero(b, list)])

def any3(a, b, c, list):
    return conde([membero(a, list)],[membero(b, list)],[membero(c, list)])

def any4(a, b, c, d, list):
    return conde([membero(a, list)],[membero(b, list)],[membero(c, list)],[membero(d, list)])

customer = var()

clues = lall(
    # Solution with 4 fields: Title, LastName, Product and Priority
    (eq, (var(), var(), var(), var()), customer),  
    # All Values 
    (membero,('Baroness', var(), var(), var()), customer),
    (membero,('Sergeant', var(), var(), var()), customer),
    (membero,('Reverend', var(), var(), var()), customer),
    (membero,('Doctor', var(), var(), var()), customer),
    (membero,(var(), 'Bevens', var(), var()), customer),
    (membero,(var(), 'Shadwell', var(), var()), customer),
    (membero,(var(), 'Dimmadome', var(), var()), customer),
    (membero,(var(), 'Rotzenheimer', var(), var()), customer),
    (membero,(var(), var(), 'Rose Bar', var()), customer),
    (membero,(var(), var(), 'Lemon Gel', var()), customer),
    (membero,(var(), var(), 'Chamomile Bar', var()), customer),
    (membero,(var(), var(), 'Hibiscus Soap-on-a-Rope', var()), customer),
    (membero,(var(), var(), var(), '1'), customer),
    (membero,(var(), var(), var(), '2'), customer),
    (membero,(var(), var(), var(), '3'), customer),
    (membero,(var(), var(), var(), '4'), customer),
    # Priority 1 > 2 > 3 > 4 
    (higher,(var(), var(), var(), '1'), (var(), var(), var(), '2'), customer), 
    (higher,(var(), var(), var(), '2'), (var(), var(), var(), '3'), customer),
    (higher,(var(), var(), var(), '3'), (var(), var(), var(), '4'), customer),
    # Clue1a  customer with the highest priority has a title and last name that begin with the same letter
    (any4,('Baroness', 'Bevens', var(), '1'),('Reverend', 'Rotzenheimer', var(), '1')
     ,('Sergeant', 'Shadwell', var(), '1'),('Doctor', 'Dimmadome', var(), '1'), customer),
    # Clue2a Bevens' priority is directly after Dimmadome
    (higher,(var(), 'Dimmadome', var(), var()), (var(), 'Bevens', var(), var()), customer),
    # Clue6 The priority of the person who ordered the Rose Bar is directly after the person who ordered the Lemon Gel.
    (higher,(var(), var(), 'Lemon Gel', var()), (var(), var(), 'Rose Bar', var()), customer),
    # Clue2b Neither of these people ordered the Chamomile Bar or the Hibiscus Soap-on-a-Rope.
    (any2, (var(), 'Dimmadome', 'Rose Bar', var()), (var(), 'Dimmadome','Lemon Gel', var()), customer),
    (any2, (var(), 'Bevens', 'Rose Bar', var()), (var(), 'Bevens', 'Lemon Gel', var()), customer),
    # Clue3 The Sergeant and the person who ordered Lemon Gel are either 1st priority or 3rd priority
    (any2, ('Sergeant', var(), var(), '1'), ('Sergeant', var(), var(), '3'), customer),
    (any2, (var(), var(), 'Lemon Gel', '1'), (var(), var(), 'Lemon Gel', '3'), customer),
    # Clue4a The Reverend didn't order the Rose Bar.
    (any3, ('Reverend', var(), 'Lemon Gel', var()), ('Reverend', var(), 'Chamomile Bar', var()),
     ('Reverend', var(), 'Hibiscus Soap-on-a-Rope', var()), customer),
    # Clue4b And the Reverend isn't 2nd priority.
    (any3, ('Reverend', var(), var(), '1'), ('Reverend', var(), var(), '3'), ('Reverend', var(), var(), '4'), customer),
    # Clue5 The Sergeant either ordered Hibiscus Soap-on-a-Rope or is 4th priority.
    (any2, ('Sergeant', var(), 'Hibiscus Soap-on-a-Rope', var()), ('Sergeant', var(), var(), '4'), customer),
    # Clue7a Dimmadome is not a Doctor
    (any3, ('Baroness', 'Dimmadome', var(), var()), ('Sergeant', 'Dimmadome', var(), var()),
     ('Reverend', 'Dimmadome', var(), var()), customer),
    # Clue7b the Baroness didn't order the Hibiscus Soap-on-a-Rope.
    (any3, ('Baroness', var(), 'Rose Bar', var()), ('Baroness', var(), 'Lemon Gel', var()),
     ('Baroness', var(), 'Chamomile Bar', var()), customer)
)    

solutions = run(0, customer, clues)

for s in solutions:
    df = pd.DataFrame(s, columns = ['Title', 'LastName', 'Product', 'Priority']) 
    df['Clue1b'] = df.apply(lambda x: not(x['Title'][0:1]==x['LastName'][0:1] and x['Priority']!='1'), axis=1)
    df = df[df['Clue1b']].drop('Clue1b',axis=1)
    if len(df)==4:
        final = df.copy()
        print(df)

File: 2020/2020W10/preppindataw10.R
library(magrittr)
library(ompr)
library(ompr.roi)
library(ROI.plugin.glpk)
library(dplyr)
library(knitr)
library(kableExtra)

n <- 4

model4 <- MILPModel() %>%
  # 6 grids:
  add_variable(xTL[i, j], i = 1:n, j = 1:n, type = "binary") %>%
  add_variable(xTX[i, j], i = 1:n, j = 1:n, type = "binary") %>%
  add_variable(xXL[i, j], i = 1:n, j = 1:n, type = "binary") %>%
  add_variable(xTP[i, j], i = 1:n, j = 1:n, type = "binary") %>%
  add_variable(xPL[i, j], i = 1:n, j = 1:n, type = "binary") %>%
  add_variable(xPX[i, j], i = 1:n, j = 1:n, type = "binary") %>%
  
  # No objective:
  set_objective(0) %>%
  
  # Only one cell can be assigned in a row:
  add_constraint(sum_expr(xTL[i, j], j = 1:n) == 1, i = 1:n) %>%
  add_constraint(sum_expr(xTX[i, j], j = 1:n) == 1, i = 1:n) %>%
  add_constraint(sum_expr(xXL[i, j], j = 1:n) == 1, i = 1:n) %>%
  add_constraint(sum_expr(xTP[i, j], j = 1:n) == 1, i = 1:n) %>%
  add_constraint(sum_expr(xPL[i, j], j = 1:n) == 1, i = 1:n) %>%
  add_constraint(sum_expr(xPX[i, j], j = 1:n) == 1, i = 1:n) %>%
  
  # Only one cell can be assigned in a column:
  add_constraint(sum_expr(xTL[i, j], i = 1:n) == 1, j = 1:n) %>%
  add_constraint(sum_expr(xTX[i, j], i = 1:n) == 1, j = 1:n) %>%
  add_constraint(sum_expr(xXL[i, j], i = 1:n) == 1, j = 1:n) %>%
  add_constraint(sum_expr(xTP[i, j], i = 1:n) == 1, j = 1:n) %>%
  add_constraint(sum_expr(xPL[i, j], i = 1:n) == 1, j = 1:n) %>%
  add_constraint(sum_expr(xPX[i, j], i = 1:n) == 1, j = 1:n) %>%
  
  #Additional
  add_constraint(xTL[i, j] + xTX[i, k] <= xXL[k, j] + 1, i = 1:n, j = 1:n, k = 1:n) %>%
  add_constraint(xTL[i, j] + xTP[i, k] <= xPL[k, j] + 1, i = 1:n, j = 1:n, k = 1:n) %>%
  add_constraint(xTX[i, j] + xTP[i, k] <= xPX[k, j] + 1, i = 1:n, j = 1:n, k = 1:n) %>%
  add_constraint(xPL[i, j] + xPX[i, k] <= xXL[k, j] + 1, i = 1:n, j = 1:n, k = 1:n)

model4_fixed <- model4 %>%
  # Clue1 Only the customer with the highest priority has a title and last name that begin with the same letter.
  add_constraint(xTL[1, 3] + xTL[2, 2] + xTL[3, 4] + xTL[4, 1] == 1) %>% 
  add_constraint(xTL[1, 3] == xTP[1, 1]) %>% 
  add_constraint(xTP[1, 1] == xPL[1, 3]) %>%
  add_constraint(xTL[2, 2] == xTP[2, 1]) %>% 
  add_constraint(xTP[2, 1] == xPL[1, 2]) %>%
  add_constraint(xTL[3, 4] == xTP[3, 1]) %>% 
  add_constraint(xTP[3, 1] == xPL[1, 4]) %>%
  add_constraint(xTL[4, 1] == xTP[4, 1]) %>% 
  add_constraint(xTP[4, 1] == xPL[1, 1]) %>%
  
  # Clue2a Bevens(L1) priority is directly after Dimmadome(L3)
  add_constraint(xPL[1, 1] == 0) %>% # Bevens(L1) cannot be 1st
  add_constraint(xPL[4, 3] == 0) %>% # Dimmadome(L3) cannot be 4th
  add_constraint(sum_expr(i * xPL[i, 1], i = 1:4) == sum_expr(i * xPL[i, 3], i = 1:4) + 1) %>% 
  # Clue2b Neither of these people(L1,L3) ordered the Chamomile Bar(X3) or the Hibiscus Soap-on-a-Rope(X4).
  add_constraint(xXL[3, 1] == 0) %>% 
  add_constraint(xXL[3, 3] == 0) %>% 
  add_constraint(xXL[4, 1] == 0) %>% 
  add_constraint(xXL[4, 3] == 0) %>% 
  
  # Clue3 The Sergeant(T2) and the person who ordered Lemon Gel(X2) are either 1st priority(P1) or 3rd priority(P3)
  add_constraint(xTP[2, 1] + xTP[2, 3] == 1) %>% 
  add_constraint(xPX[1, 2] + xPX[3, 2] == 1) %>% 
  
  # Clue4 The Reverend(T3) didn't order the Rose Bar(X1) and isn't 2nd priority(P2).
  add_constraint(xTX[3, 1] == 0) %>% 
  add_constraint(xTP[3, 2] == 0) %>% 
  
  # Clue5 The Sergeant(T2) either ordered Hibiscus Soap-on-a-Rope(X4) or is 4th priority(P4)
  add_constraint(xTX[2, 4] + xTP[2, 4] >= 1) %>% 
  
  # Clue6 The priority of the person who ordered the Rose Bar(X1) is directly after the person who ordered the Lemon Gel(X2).
  add_constraint(xPX[1, 1] == 0) %>% # Rose Bar(X1) cannot be 1st(P1)
  add_constraint(xPX[4, 2] == 0) %>% # Lemon Gel(X2) cannot be 4th(P4)
  add_constraint(sum_expr(i * xPX[i, 1], i = 1:4) == sum_expr(i * xPX[i, 2], i = 1:4) + 1) %>%
  
  # Clue7 Dimmadome(L3) is not a Doctor(T1) and the Baroness(T4) didn't order the Hibiscus Soap-on-a-Rope(X4).
  add_constraint(xTL[1, 3] == 0) %>%
  add_constraint(xTX[4, 4] == 0) 
  
  

result <- solve_model(model4_fixed, with_ROI(solver = "glpk", verbose = TRUE))

Is <- c("Doctor", "Sergeant", "Reverend", "Baroness") # Title
Js <- c("Bevens", "Shadwell", "Dimmadome", "Rotzenheimer") # LastName
Ks <- c("Rose Bar", "Lemon Gel", "Chamomile Bar", "Hibiscus Soap-on-a-Rope") # Product
Ls <- as.character(1:4) # Priority

# Extract solved grids
solution <- list(
  xTL = get_solution(result, xTL[i, j]) %>%
    dplyr::filter(value == 1) %>%
    dplyr::select(-c(variable, value)) %>%
    dplyr::mutate(i = Is[i], j = Js[j]),
  xTX = get_solution(result, xTX[i, k]) %>%
    dplyr::filter(value == 1) %>%
    dplyr::select(-c(variable, value)) %>%
    dplyr::mutate(i = Is[i], k = Ks[k]),
  xTP = get_solution(result, xTP[i, l]) %>%
    dplyr::filter(value == 1) %>%
    dplyr::select(-c(variable, value)) %>%
    dplyr::mutate(i = Is[i], l = Ls[l]),
  xPL = get_solution(result, xPL[l, j]) %>%
    dplyr::filter(value == 1) %>%
    dplyr::select(-c(variable, value)) %>%
    dplyr::mutate(l = Ls[l], j = Js[j]),
  xPX = get_solution(result, xPX[l, k]) %>%
    dplyr::filter(value == 1) %>%
    dplyr::select(-c(variable, value)) %>%
    dplyr::mutate(l = Ls[l], k = Ks[k]),
  xXL = get_solution(result, xXL[k, j]) %>%
    dplyr::filter(value == 1) %>%
    dplyr::select(-c(variable, value)) %>%
    dplyr::mutate(k = Ks[k], j = Js[j])
)

# Obtained solution
solution %>%
  purrr::reduce(dplyr::left_join) %>%
  dplyr::arrange(l) %>%
  kable(col.names = c("Title", "LastName", "Product", "Priority"), align = "l") %>%
  kable_styling(bootstrap_options = c("striped"))